bpo-32236: Issue RuntimeWarning if buffering=1 for open() in binary mode

izbyshev opened the pull request:
If buffering=1 is specified for open() in binary mode, it is silently
treated as buffering=-1 (i.e., the default buffer size).
Coupled with the fact that line buffering is always supported in Python 2,
such behavior caused several issues (e.g., bpo-10344, bpo-21332).

Raise ValueError on attempts to use line buffering in binary mode to prevent
possible bugs and expose existing buggy code.

<!-- issue-number: bpo-32236 -->
https://bugs.python.org/issue32236
<!-- /issue-number -->

Commit:
bpo-32236: Don't allow buffering=1 in open() for binary mode

If buffering=1 is specified for open() in binary mode, it is silently
treated as buffering=-1 (i.e., the default buffer size).
Coupled with the fact that line buffering is always supported in Python 2,
such behavior caused several issues (e.g., bpo-10344, bpo-21332).

Raise ValueError on attempts to use line buffering in binary mode to prevent
possible bugs and expose existing buggy code.

Commit:
bpo-32236: Issue RuntimeWarning instead of raising ValueError

gpshead reviewed (approved):
The subprocess related changes are correct, others should review the other changes.  Though at first glance I believe those to also be correct.

vstinner commented on Lib/test/test_file.py:
```diff
@@ -164,22 +164,31 @@ def testBadModeArgument(self):
+        try:
+            f = self.open(TESTFN, 'wb', s)
+            f.write(str(s).encode("ascii"))
+            f.close()
```
what's the point of calling close() twice here? with should be used, no?

pitrou replied:
This is just a refactor of an existing test.

vstinner commented on Lib/test/test_file.py:
```diff
@@ -164,22 +164,31 @@ def testBadModeArgument(self):
-            except OSError as msg:
-                self.fail('error setting buffer size %d: %s' % (s, str(msg)))
-            self.assertEqual(d, s)
+        for s in (-1, 0, 512):
```
Would it be possible to check that these calls don't emit RuntimeWarning?

izbyshev replied:
I've added the check.

vstinner reviewed (changes requested):
Please add a NEWS entry to mention that open() now emits a warning if the file is opened in binary mode with buffering=1. Moreover, the NEWS entry should mention that the change fixes a performance issue in subprocess.Popen if buffering=1: the underlying binary file is no longer opened with inefficient buffering=1, if I understood correctly.

pitrou commented on Lib/_pyio.py:
```diff
@@ -198,6 +198,10 @@ def open(file, mode="r", buffering=-1, encoding=None, errors=None,
         raise ValueError("binary mode doesn't take a newline argument")
+    if binary and buffering == 1:
+        import warnings
+        warnings.warn("line buffering is not supported in binary mode",
```
It's not necessary obvious why the warning talks about line buffering, so I would be a bit more explicit: `"line buffering (buffering = 1) isn't supported in binary mode, file will be fully buffered."`

izbyshev replied:
OK, I've clarified the message as you suggest.

pitrou commented on Lib/test/test_io.py:
```diff
@@ -591,15 +591,23 @@ def test_large_file_ops(self):
             self.large_file_ops(f)
 
+    def _checked_open(self, path, mode, bufsize, under_check=False):
+        if not under_check and bufsize == 1:
```
What is `under_check` for? It doesn't seem to be ever True.

izbyshev replied:
This function was a conditional context manager, and `under_check` is set to `True` in the recursive call. But I've removed that function anyway to have only one place with warning check (in test_file.py).

pitrou commented on Lib/test/test_file.py:
```diff
@@ -164,22 +164,31 @@ def testBadModeArgument(self):
+        for s in (-1, 0, 512):
+            self._checkBufferSize(s)
+
+        # test that attempts to use line buffering in binary mode cause
```
No need to test this both here and in `test_io.py`, IMHO.

izbyshev replied:
OK, I've preserved only one check here. I think that `test_with_open` in `test_io.py` wasn't a really good place for that check after all.

pitrou reviewed (dismissed):
Thanks. You need to add a NEWS file using the [blurb](https://pypi.org/project/blurb/) utility. Also, see comments below.

Commit:
Clarify warning message

Commit:
Remove RuntimeWarning test from test_io.py

Commit:
Fix check for warning message

Commit:
Add NEWS entry

Commit:
Check that no warning is emitted when buffering != 1

Commit:
Fix warning message check in test_subprocess.py

Commit:
Update docs for codecs.open()

Commit:
Fix check_no_warnings() docs to reflect what it actually does

My previous understanding of behavior of check_no_resource_warning(),
which check_no_warnings() is based on, was incorrect.

vstinner commented on Lib/_pyio.py:
```diff
@@ -198,6 +198,10 @@ def open(file, mode="r", buffering=-1, encoding=None, errors=None,
+    if binary and buffering == 1:
+        import warnings
+        warnings.warn("line buffering (buffering=1) isn't supported in binary "
+                      "mode, file will be fully buffered", RuntimeWarning, 2)
```
There is no such thing as "fully buffered". I propose to also say "the default size will be used" here.

Extract of the code:

```
        if buffering == 1 or buffering < 0 and raw.isatty():
            buffering = -1
            line_buffering = True
        if buffering < 0:
            buffering = DEFAULT_BUFFER_SIZE
```

buffering=1 becomes buffering=DEFAULT_BUFFER_SIZE if I understand correctly.

izbyshev replied:
Fixed, thanks!

vstinner commented on Lib/subprocess.py:
```diff
@@ -721,12 +721,20 @@ def __init__(self, args, bufsize=-1, executable=None,
+        if self.text_mode:
+            if bufsize == 1:
+                line_buffering = True
+                # line buffering is not supported in binary mode
```
But here we are in text mode, no? The comment is misleading, I suggest to remove it.

Or say something like: "# Use the default buffer size for the binary buffered reader/writer".

izbyshev replied:
I was referring to the binary mode used for the underlying streams, and I didn't realize that it can be misleading, but now I see it. Changed to clearer wording, thanks!

vstinner commented on Modules/_io/_iomodule.c:
```diff
@@ -362,6 +362,13 @@ _io_open_impl(PyObject *module, PyObject *file, const char *mode,
+    if (binary && buffering == 1) {
+        if (PyErr_WarnEx(PyExc_RuntimeWarning,
+                         "line buffering (buffering=1) isn't supported in "
+                         "binary mode, file will be fully buffered", 1) < 0)
```
Ditto: say something like "the default size will be used".

PEP 7: please add { ... } around the "goto error;". By the way, see https://nakedsecurity.sophos.com/2014/02/24/anatomy-of-a-goto-fail-apples-ssl-bug-explained-plus-an-unofficial-patch/ :-)

izbyshev replied:
Yes, I'm aware of PEP 7 and `goto fail` bug, but I usually follow the rule of local consistency, and a similar code above doesn't use brackets. Now I've added brackets here as you suggested.

vstinner commented on Misc/NEWS.d/next/Core and Builtins/2018-09-11-23-50-40.bpo-32236.3RupnN.rst:
```diff
@@ -0,0 +1,2 @@
+Warn that line buffering is not supported if :func:`open` is called with
+binary mode and ``buffering=1``.
```
I would prefer to be more explicit and list more impact functions: open, io.open, codecs.open. Would it be possible to put functions first in the sentence? Something like: "open(), io.open(), codecs.open() now emit a warning if the file is opened in binary mode with buffering=1." So the reader directly gets the context.

izbyshev replied:
Some other functions are impacted too, like `pathlib.Path.open()` and `tempfile.TemporaryFile()` (and some other classes from `tempfile`). Do you think we should find and list all such functions here? Maybe we should just provide a couple of examples: `io.open` and the standard library functions relying on it (e.g., `codecs.open`, `pathlib.Path.open`) now issue a warning if a file is opened in binary mode with `buffering=1`.

vstinner reviewed (changes requested):
It is now way better, but I still have a few comments.

Commit:
Clarify warning message

Commit:
Clarify a comment
